File: plugins/devkit/scripts/check-dig-migration.mjs
#!/usr/bin/env node
import fs from "fs";
import path from "path";
import os from "os";
import crypto from "crypto";
import { execFileSync } from "child_process";

const args = process.argv.slice(2);
const modeArg = args.find((a) => a.startsWith("--mode="));
const artifactArg = args.find((a) => a.startsWith("--artifact="));
const mode = modeArg ? modeArg.split("=")[1] : "repo";
const artifactPath = artifactArg ? artifactArg.split("=")[1] : "";

const root = process.cwd();
const legacyPatterns = [
  /\/devkit:codex(?!-)\b/g,
  /\/devkit:agent-orch-(core|openai|anthropic|google)\b/g,
  /devkit-codex(?!-)\b/g,
  /devkit-agent-orch-core\b/g,
  /agent-orch-(core|openai|anthropic|google)\b/g,
  /plugins\/devkit\/skills\/codex\b/g,
  /plugins\/devkit\/skills\/agent-orch-(core|openai|anthropic|google)\b/g,
];

const textExt = new Set([
  ".md", ".txt", ".json", ".yaml", ".yml", ".js", ".mjs", ".cjs", ".ts", ".tsx", ".jsx", ".sh", ".ps1", ".bat", ".cmd", ".toml", ".ini", ".cfg", ".py"
]);

function isBinaryBuffer(buf) {
  const max = Math.min(buf.length, 8000);
  for (let i = 0; i < max; i += 1) {
    if (buf[i] === 0) return true;
  }
  return false;
}

function collectFiles(dir, out = []) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const e of entries) {
    if (e.name === ".git" || e.name === "node_modules" || e.name === ".claude" || e.name === ".codex") continue;
    const p = path.join(dir, e.name);
    if (e.isDirectory()) collectFiles(p, out);
    else out.push(p);
  }
  return out;
}

function isAllowedException(fileRel, line, readmeSectionState) {
  if (line.includes("migration-allow")) return true;
  if (fileRel === "CHANGELOG.md") return true;
  if (fileRel === "README.md") {
    return readmeSectionState.inMigrationNotice;
  }
  return false;
}

function scanTextFile(absPath, relPath) {
  const content = fs.readFileSync(absPath, "utf8");
  const lines = content.split(/\r?\n/);
  const findings = [];
  const readmeState = { inMigrationNotice: false };

  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i];

    if (relPath === "README.md") {
      if (/^##\s+Migration Notice\b/.test(line)) {
        readmeState.inMigrationNotice = true;
      } else if (/^##\s+/.test(line) && readmeState.inMigrationNotice) {
        readmeState.inMigrationNotice = false;
      }
    }

    for (const pat of legacyPatterns) {
      const m = line.match(pat);
      if (!m) continue;
      if (isAllowedException(relPath, line, readmeState)) continue;
      findings.push({
        path: relPath,
        line: i + 1,
        token: m[0],
        replacement: "Use /dig runtime adapters and new templates",
      });
    }
  }

  return findings;
}

function scanDir(dir) {
  const findings = [];
  const binaries = [];
  const files = collectFiles(dir);

  for (const file of files) {
    const rel = path.relative(root, file).replace(/\\/g, "/");
    if (rel.startsWith("plugins/devkit/scripts/check-dig-migration.mjs")) continue;
    if (rel.startsWith("plugins/devkit/scripts/check-skill-surface.mjs")) continue;
    if (rel.startsWith("plugins/devkit/scripts/check-dig-routing.mjs")) continue;
    const ext = path.extname(file).toLowerCase();

    const buf = fs.readFileSync(file);
    if (isBinaryBuffer(buf) || !textExt.has(ext)) {
      const hash = crypto.createHash("sha256").update(buf).digest("hex");
      binaries.push({ path: rel, sha256: hash });
      continue;
    }

    findings.push(...scanTextFile(file, rel));
  }

  return { findings, binaries };
}

function extractArtifact(artifact, tempDir) {
  const lower = artifact.toLowerCase();
  if (lower.endsWith(".zip")) {
    if (process.platform === "win32") {
      execFileSync("powershell", ["-NoProfile", "-Command", `Expand-Archive -Path \"${artifact}\" -DestinationPath \"${tempDir}\" -Force`], { stdio: "inherit" });
    } else {
      execFileSync("unzip", ["-o", artifact, "-d", tempDir], { stdio: "inherit" });
    }
    return;
  }
  if (lower.endsWith(".tar") || lower.endsWith(".tgz") || lower.endsWith(".tar.gz")) {
    execFileSync("tar", ["-xf", artifact, "-C", tempDir], { stdio: "inherit" });
    return;
  }
  throw new Error(`Unsupported artifact type: ${artifact}`);
}

function main() {
  let result;

  if (mode === "repo") {
    result = scanDir(root);
  } else if (mode === "artifact") {
    if (!artifactPath) throw new Error("--artifact is required in artifact mode");
    const absArtifact = path.resolve(root, artifactPath);
    if (!fs.existsSync(absArtifact)) throw new Error(`Artifact not found: ${absArtifact}`);
    const tmp = fs.mkdtempSync(path.join(os.tmpdir(), "dig-mig-"));
    extractArtifact(absArtifact, tmp);
    result = scanDir(tmp);
  } else {
    throw new Error(`Unknown mode: ${mode}`);
  }

  const out = {
    mode,
    findings: result.findings,
    binaryEntries: result.binaries,
  };

  console.log(JSON.stringify(out, null, 2));

  if (result.findings.length > 0) {
    process.exit(2);
  }
}

main();

File: plugins/devkit/scripts/check_dig_migration.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import hashlib
import json
import tarfile
import zipfile
from pathlib import Path
from tempfile import TemporaryDirectory


ROOT = Path.cwd()
LEGACY_PATTERNS = [
    r"/devkit:dig\b",
    r"/prompts:devkit-dig\b",
    r"/devkit:codex(?!-)\b",
    r"/devkit:agent-orch-(core|openai|anthropic|google)\b",
    r"devkit-codex(?!-)\b",
    r"devkit-agent-orch-core\b",
    r"agent-orch-(core|openai|anthropic|google)\b",
    r"plugins/devkit/skills/codex\b",
    r"plugins/devkit/skills/agent-orch-(core|openai|anthropic|google)\b",
]
TEXT_EXT = {
    ".md",
    ".txt",
    ".lock",
    ".json",
    ".js",
    ".mjs",
    ".cjs",
    ".yaml",
    ".yml",
    ".ts",
    ".tsx",
    ".jsx",
    ".sh",
    ".ps1",
    ".bat",
    ".cmd",
    ".toml",
    ".ini",
    ".cfg",
    ".py",
}


def is_binary_buffer(buf: bytes) -> bool:
    return b"\x00" in buf[:8000]


def collect_files(directory: Path) -> list[Path]:
    out: list[Path] = []
    for path in directory.rglob("*"):
        if path.is_dir():
            continue
        rel = path.relative_to(directory).as_posix()
        if any(part in {".git", "node_modules", ".claude", ".codex", ".venv", "__pycache__"} for part in path.parts):
            continue
        out.append(path)
    return out


def is_allowed_exception(file_rel: str, line: str, in_migration_notice: bool) -> bool:
    if "migration-allow" in line:
        return True
    if file_rel == "CHANGELOG.md":
        return True
    if file_rel == "README.md":
        return in_migration_notice
    return False


def scan_text_file(abs_path: Path, rel_path: str) -> list[dict[str, object]]:
    import re

    content = abs_path.read_text(encoding="utf-8")
    lines = content.splitlines()
    findings: list[dict[str, object]] = []
    in_migration_notice = False

    for index, line in enumerate(lines, start=1):
        if rel_path == "README.md":
            if re.match(r"^##\s+Migration Notice\b", line):
                in_migration_notice = True
            elif re.match(r"^##\s+", line) and in_migration_notice:
                in_migration_notice = False

        for pattern in LEGACY_PATTERNS:
            match = re.search(pattern, line)
            if not match:
                continue
            if is_allowed_exception(rel_path, line, in_migration_notice):
                continue
            findings.append(
                {
                    "path": rel_path,
                    "line": index,
                    "token": match.group(0),
                    "replacement": "Use /dig runtime adapters and new templates",
                }
            )

    return findings


def scan_dir(directory: Path) -> dict[str, list[dict[str, object]]]:
    findings: list[dict[str, object]] = []
    binaries: list[dict[str, object]] = []

    for file_path in collect_files(directory):
        rel = file_path.relative_to(directory).as_posix()
        if rel.startswith("plugins/devkit/scripts/check_dig_migration.py"):
            continue
        if rel.startswith("plugins/devkit/scripts/check_skill_surface.py"):
            continue
        if rel.startswith("plugins/devkit/scripts/check_dig_routing.py"):
            continue
        if rel.startswith("plugins/devkit/scripts/check_plugin_version_bump.py"):
            continue

        buf = file_path.read_bytes()
        if is_binary_buffer(buf) or file_path.suffix.lower() not in TEXT_EXT:
            binaries.append({"path": rel, "sha256": hashlib.sha256(buf).hexdigest()})
            continue

        findings.extend(scan_text_file(file_path, rel))

    return {"findings": findings, "binaries": binaries}


def extract_artifact(artifact: Path, temp_dir: Path) -> None:
    lower = artifact.name.lower()
    if lower.endswith(".zip"):
        with zipfile.ZipFile(artifact) as archive:
            archive.extractall(temp_dir)
        return
    if lower.endswith((".tar", ".tgz", ".tar.gz")):
        with tarfile.open(artifact) as archive:
            archive.extractall(temp_dir)
        return
    raise RuntimeError(f"Unsupported artifact type: {artifact}")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", default="repo", choices=["repo", "artifact"])
    parser.add_argument("--artifact", default="")
    args = parser.parse_args()

    if args.mode == "repo":
        result = scan_dir(ROOT)
    else:
        if not args.artifact:
            raise RuntimeError("--artifact is required in artifact mode")
        artifact = (ROOT / args.artifact).resolve()
        if not artifact.exists():
            raise RuntimeError(f"Artifact not found: {artifact}")
        with TemporaryDirectory(prefix="dig-mig-") as temp_dir_name:
            temp_dir = Path(temp_dir_name)
            extract_artifact(artifact, temp_dir)
            result = scan_dir(temp_dir)

    payload = {
        "mode": args.mode,
        "findings": result["findings"],
        "binaryEntries": result["binaries"],
    }
    print(json.dumps(payload, ensure_ascii=False, indent=2))
    return 2 if result["findings"] else 0


if __name__ == "__main__":
    raise SystemExit(main())
